src/uWS_Client.cpp

```diff
@@ -0,0 +1,1371 @@
+        int serverMaxWindowBits = 0;
+        int clientMaxWindowBits = 0;
+
+        NegotiationOffer(char *in)
+        {
+            int token = 1;
+            for (; token && token != PERMESSAGE_DEFLATE; token = getToken(&in));
+
+            while ((token = getToken(&in))) {
+                switch (token) {
+                case PERMESSAGE_DEFLATE:
+                    return;
+                case SERVER_NO_CONTEXT_TAKEOVER:
+                    serverNoContextTakeover = true;
+                    break;
+                case CLIENT_NO_CONTEXT_TAKEOVER:
+                    clientNoContextTakeover = true;
+                    break;
+                case SERVER_MAX_WINDOW_BITS:
+                    serverMaxWindowBits = 1;
+                    lastInteger = &serverMaxWindowBits;
+                    break;
+                case CLIENT_MAX_WINDOW_BITS:
+                    clientMaxWindowBits = 1;
+                    lastInteger = &clientMaxWindowBits;
+                    break;
+                default:
+                    if (token < 0 && lastInteger) {
+                        *lastInteger = -token;
+                    } else {
+                        cout << "UNKNOWN TOKEN: " << token << endl;
+                    }
+                    break;
+                }
+            }
+        }
+    };
+
+    PerMessageDeflate(string &request, int options, string &response) : readStream({}), writeStream({})
+    {
+        NegotiationOffer offer((char *) request.c_str());
+
+        response = "Sec-WebSocket-Extensions: permessage-deflate";
+        if ((options & SERVER_NO_CONTEXT_TAKEOVER) || offer.serverNoContextTakeover) {
+            response += "; server_no_context_takeover";
+            serverNoContextTakeover = true;
+        }
+        if ((options & CLIENT_NO_CONTEXT_TAKEOVER) || offer.clientNoContextTakeover) {
+            response += "; client_no_context_takeover";
+            clientNoContextTakeover = true;
+        }
+
+        inflateInit2(&readStream, -15);
+    }
+
+    ~PerMessageDeflate()
+    {
+        inflateEnd(&readStream);
+    }
+
+    bool inflate(char *src, size_t srcLength, char *dst, size_t &dstLength)
+    {
+        readStream.next_in = (unsigned char *) src;
+        readStream.avail_in = srcLength;
+        readStream.next_out = (unsigned char *) dst;
+        readStream.avail_out = dstLength;
+
+        if (Z_OK != ::inflate(&readStream, Z_NO_FLUSH)) {
+            dstLength = 0;
+            return true;
+        } else {
+            dstLength = (dstLength - readStream.avail_out);
+            return false;
+        }
+    }
+};
+
+struct SocketData {
+    unsigned char state = READ_HEAD;
+    unsigned char sendState = FRAGMENT_START;
+    unsigned char fin = true;
+    char opStack = -1;
+    char spill[16]; // can be 14 in size
+    unsigned char spillLength = 0;
+    OpCode opCode[2];
+    unsigned int remainingBytes = 0;
+    char mask[4];
+    Client *client;
```

Comment:
You don't need SocketData -> Client relation, Client knows its SocketData. You do need some kind of shared buffer holder though (receiveBuffer, sendBuffer, inflateBuffer should not be per client but shared among all clients just like the Server does)